[Bug] Impossible to include "wasmtime.hh" multiple times (multiple definition errors at linking)?

Hello,

being new to WebAssembly, i tried to implement a WASM-based plugin-concept. The wasmtime project looks awesome and a first hello-world like example (with some additions like using a flatbuffers-based plugin-api to ease non-fundamental data-type passing) worked great.

Then i tried to flesh out the example to something bigger.
**In it's essence, we are just talking about increasing the cardinality of compilation-units here**.

Without being an expert C++ developer, **i carefully ask if it's possible, that the current implementation is broken in some sense as we cannot include "wasmtime.hh" multiple times without violating the one-definition rule?** 

My problem exactly looks like [this StackOverflow question](https://stackoverflow.com/a/29196891/2320035) (please don't be offended by the negative tone there).

Here is a minimal example (which does not make sense in this state) which should fail for you too (i guess):

# Problem Demo

### `run.cc`

    #include <iostream>
    
    #include "wasm_backend.hh"
    
    int main() {
      std::cout << "hello" << std::endl; 
      return 0;
    }

### `wasm_backend.hh`

    #pragma once // same with "classic" include-guards
   
    // This include leads to linking-errors 
    #include <wasmtime.hh>
  
    class WasmBackend {};

### `wasm_backend.cc`

    #include "wasm_backend.hh"

    // no other code here

# Compilation

    clang++ run.cc wasm_backend.cc -std=c++20 -I third_party/wasmtime-v0.28.0-x86_64-linux-c-api/include third_party/wasmtime-v0.28.0-x86_64-linux-c-api/lib/libwasmtime.a -lpthread -ldl -lm -o run

# Error
    
    sascha@venezia:~/Dev/wasm_plugin_concept_demo/bugreport$ clang++ run.cc wasm_backend.cc -std=c++20 -I third_party/wasmtime-v0.28.0-x86_64-linux-c-api/include third_party/wasmtime-v0.28.0-x86_64-linux-c-api/lib/libwasmtime.a -lpthread -ldl -lm -o run
    /usr/bin/ld: /tmp/wasm_backend-8e2203.o: in function `wasmtime::Val::Val(wasmtime::Func)':
    wasm_backend.cc:(.text+0x310): multiple definition of `wasmtime::Val::Val(wasmtime::Func)'; /tmp/run-b91316.o:run.cc:(.text+0x310): first defined here
    /usr/bin/ld: /tmp/wasm_backend-8e2203.o: in function `wasmtime::Val::Val(wasmtime::ExternRef)':
    wasm_backend.cc:(.text+0x370): multiple definition of `wasmtime::Val::Val(wasmtime::ExternRef)'; /tmp/run-b91316.o:run.cc:(.text+0x370): first defined here
    /usr/bin/ld: /tmp/wasm_backend-8e2203.o: in function `wasmtime::Val::Val(std::optional<wasmtime::Func>)':
    wasm_backend.cc:(.text+0x290): multiple definition of `wasmtime::Val::Val(std::optional<wasmtime::Func>)'; /tmp/run-b91316.o:run.cc:(.text+0x290): first defined here
    /usr/bin/ld: /tmp/wasm_backend-8e2203.o: in function `wasmtime::Val::Val(wasmtime::Func)':
    wasm_backend.cc:(.text+0x310): multiple definition of `wasmtime::Val::Val(wasmtime::Func)'; /tmp/run-b91316.o:run.cc:(.text+0x310): first defined here
    /usr/bin/ld: /tmp/wasm_backend-8e2203.o: in function `wasmtime::Val::Val(wasmtime::ExternRef)':
    wasm_backend.cc:(.text+0x370): multiple definition of `wasmtime::Val::Val(wasmtime::ExternRef)'; /tmp/run-b91316.o:run.cc:(.text+0x370): first defined here
    /usr/bin/ld: /tmp/wasm_backend-8e2203.o: in function `wasmtime::Val::Val(std::optional<wasmtime::Func>)':
    wasm_backend.cc:(.text+0x290): multiple definition of `wasmtime::Val::Val(std::optional<wasmtime::Func>)'; /tmp/run-b91316.o:run.cc:(.text+0x290): first defined here
    /usr/bin/ld: /tmp/wasm_backend-8e2203.o: in function `wasmtime::Store::Context::Context(wasmtime::Caller*)':
    wasm_backend.cc:(.text+0x260): multiple definition of `wasmtime::Store::Context::Context(wasmtime::Caller*)'; /tmp/run-b91316.o:run.cc:(.text+0x260): first defined here
    /usr/bin/ld: /tmp/wasm_backend-8e2203.o: in function `wasmtime::Store::Context::Context(wasmtime::Caller&)':
    wasm_backend.cc:(.text+0x220): multiple definition of `wasmtime::Store::Context::Context(wasmtime::Caller&)'; /tmp/run-b91316.o:run.cc:(.text+0x220): first defined here
    /usr/bin/ld: /tmp/wasm_backend-8e2203.o: in function `wasmtime::Store::Context::Context(wasmtime::Caller*)':
    wasm_backend.cc:(.text+0x260): multiple definition of `wasmtime::Store::Context::Context(wasmtime::Caller*)'; /tmp/run-b91316.o:run.cc:(.text+0x260): first defined here
    /usr/bin/ld: /tmp/wasm_backend-8e2203.o: in function `wasmtime::Store::Context::Context(wasmtime::Caller&)':
    wasm_backend.cc:(.text+0x220): multiple definition of `wasmtime::Store::Context::Context(wasmtime::Caller&)'; /tmp/run-b91316.o:run.cc:(.text+0x220): first defined here
    /usr/bin/ld: /tmp/wasm_backend-8e2203.o: in function `wasmtime::Caller::get_export(std::basic_string_view<char, std::char_traits<char> >)':
    wasm_backend.cc:(.text+0x460): multiple definition of `wasmtime::Caller::get_export(std::basic_string_view<char, std::char_traits<char> >)'; /tmp/run-b91316.o:run.cc:(.text+0x460): first defined here
    /usr/bin/ld: /tmp/wasm_backend-8e2203.o: in function `wasmtime::wat2wasm(std::basic_string_view<char, std::char_traits<char> >)':
    wasm_backend.cc:(.text+0x50): multiple definition of `wasmtime::wat2wasm(std::basic_string_view<char, std::char_traits<char> >)'; /tmp/run-b91316.o:run.cc:(.text+0x50): first defined here
    /usr/bin/ld: /tmp/wasm_backend-8e2203.o: in function `wasmtime::operator<<(std::ostream&, wasmtime::Error const&)':
    wasm_backend.cc:(.text+0x0): multiple definition of `wasmtime::operator<<(std::ostream&, wasmtime::Error const&)'; /tmp/run-b91316.o:run.cc:(.text+0x0): first defined here
    /usr/bin/ld: /tmp/wasm_backend-8e2203.o: in function `wasmtime::Val::funcref() const':
    wasm_backend.cc:(.text+0x3d0): multiple definition of `wasmtime::Val::funcref() const'; /tmp/run-b91316.o:run.cc:(.text+0x3d0): first defined here

# Description
We just *start* to wrap a `wasm_backend` which *includes* `wasmtime.hh` and is *included* in it's sources AND in one other source (run.cc / main here).  

The test-cases are not catching a problem like that it seems.

Thanks for all your work!
- Sascha

--- a/include/wasmtime.hh
+++ b/include/wasmtime.hh
@@ -133,7 +133,7 @@ public:
 };
 
 /// \brief Used to print an error.
-std::ostream &operator<<(std::ostream &os, const Error &e) {
+inline std::ostream &operator<<(std::ostream &os, const Error &e) {
   os << e.message();
   return os;
 }
@@ -424,7 +424,8 @@ public:
  *
  * Returns either an error if parsing failed or the wasm binary.
  */
-[[nodiscard]] Result<std::vector<uint8_t>> wat2wasm(std::string_view wat) {
+[[nodiscard]] inline Result<std::vector<uint8_t>>
+wat2wasm(std::string_view wat) {
   wasm_byte_vec_t ret;
   auto *error = wasmtime_wat2wasm(wat.data(), wat.size(), &ret);
   if (error != nullptr) {
@@ -2009,9 +2010,9 @@ public:
   Store::Context context() { return this; }
 };
 
-Store::Context::Context(Caller &caller)
+inline Store::Context::Context(Caller &caller)
     : Context(wasmtime_caller_context(caller.ptr)) {}
-Store::Context::Context(Caller *caller) : Context(*caller) {}
+inline Store::Context::Context(Caller *caller) : Context(*caller) {}
 
 /**
  * \brief Representation of a WebAssembly function.
@@ -2140,7 +2141,7 @@ public:
   }
 };
 
-Val::Val(std::optional<Func> func) : val{} {
+inline Val::Val(std::optional<Func> func) : val{} {
   val.kind = WASMTIME_FUNCREF;
   val.of.funcref.store_id = 0;
   val.of.funcref.index = 0;
@@ -2149,10 +2150,10 @@ Val::Val(std::optional<Func> func) : val{} {
   }
 }
 
-Val::Val(Func func) : Val(std::optional(func)) {}
-Val::Val(ExternRef ptr) : Val(std::optional(ptr)) {}
+inline Val::Val(Func func) : Val(std::optional(func)) {}
+inline Val::Val(ExternRef ptr) : Val(std::optional(ptr)) {}
 
-std::optional<Func> Val::funcref() const {
+inline std::optional<Func> Val::funcref() const {
   if (val.kind != WASMTIME_FUNCREF) {
     std::abort();
   }
@@ -2527,7 +2528,7 @@ public:
   }
 };
 
-std::optional<Extern> Caller::get_export(std::string_view name) {
+inline std::optional<Extern> Caller::get_export(std::string_view name) {
   wasmtime_extern_t item;
   if (wasmtime_caller_export_get(ptr, name.data(), name.size(), &item)) {
     return Instance::cvt(item);

--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -10,12 +10,17 @@ FetchContent_MakeAvailable(googletest)
 include(GoogleTest)
 
 function(add_test name)
-
   add_executable(test-${name} ${name}.cc)
   target_link_libraries(test-${name} PRIVATE wasmtime-cpp gtest_main)
-
   gtest_discover_tests(test-${name})
 endfunction()
 
 add_test(simple)
 add_test(types)
+
+# Add a custom test where two files include `wasmtime.hh` and are compiled into
+# the same executable (basically makes sure any defined functions in the header
+# are tagged with `inline`).
+add_executable(test-double-include double-include-a.cc double-include-b.cc)
+target_link_libraries(test-double-include PRIVATE wasmtime-cpp gtest_main)
+gtest_discover_tests(test-double-include)

--- a/tests/double-include-a.cc
+++ b/tests/double-include-a.cc
@@ -1,0 +1,9 @@
+#include <gtest/gtest.h>
+#include <wasmtime.hh>
+
+using namespace wasmtime;
+
+TEST(Store, WorksInFileA) {
+  Engine engine;
+  Store store(engine);
+}

--- a/tests/double-include-b.cc
+++ b/tests/double-include-b.cc
@@ -1,0 +1,9 @@
+#include <gtest/gtest.h>
+#include <wasmtime.hh>
+
+using namespace wasmtime;
+
+TEST(Store, WorksInFileB) {
+  Engine engine;
+  Store store(engine);
+}

